File: SortVision/public/code/selection/c/selectionSort.c
#include <stdio.h>

// Function to find the index of the minimum element in the array from start to end
int findMinIndex(int arr[], int start, int end) {
    int minIndex = start;
    for (int i = start + 1; i <= end; i++) {
        if (arr[i] < arr[minIndex]) {
            minIndex = i;
        }
    }
    return minIndex;
}

// Function to perform Selection Sort
void selectionSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        // Find the index of the minimum element in the unsorted part
        int minIndex = findMinIndex(arr, i, n - 1);

        // Swap the found minimum element with the current element
        if (minIndex != i) {
            int temp = arr[i];
            arr[i] = arr[minIndex];
            arr[minIndex] = temp;
        }
    }
}

// Function to print the array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    // Example usage with test cases
    int arr1[] = {64, 25, 12, 22, 11};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);

    printf("Original array: ");
    printArray(arr1, n1);

    selectionSort(arr1, n1);

    printf("Sorted array: ");
    printArray(arr1, n1);

    // Test case: Single element
    int arr2[] = {42};
    printf("\nSingle element array before: ");
    printArray(arr2, 1);
    selectionSort(arr2, 1);
    printf("After sorting: ");
    printArray(arr2, 1);

    // Test case: Empty array
    int arr3[] = {};
    printf("\nEmpty array before: ");
    printArray(arr3, 0);
    selectionSort(arr3, 0);
    printf("After sorting: ");
    printArray(arr3, 0);

    return 0;
}

/*
📌 Time Complexity:
- Worst Case: O(n^2)
- Average Case: O(n^2)
- Best Case: O(n^2) (even if already sorted)

📌 Space Complexity:
- O(1) auxiliary space (in-place sort)

📌 Notes:
- Not stable by default
- Good for small datasets
- Minimizes number of swaps compared to Bubble Sort

📌 Edge Cases Handled:
- Empty array
- Single element array
*/

File: SortVision/public/code/selection/r/selectionSort.r
# selectionSort.r
# ----------------------------
# 📊 Implementation of Selection Sort in R
# ----------------------------------------

#' Find the index of the minimum element in a subvector
#'
#' @param arr A numeric vector
#' @param start_idx Start index of the unsorted subvector
#' @param end_idx End index of the vector
#' @return Index of the minimum element between start_idx and end_idx
find_min_index <- function(arr, start_idx, end_idx) {
  min_idx <- start_idx
  for (i in (start_idx + 1):end_idx) {
    if (arr[i] < arr[min_idx]) {
      min_idx <- i
    }
  }
  return(min_idx)
}

#' Perform Selection Sort on a numeric vector
#'
#' @param arr A numeric vector to be sorted
#' @return A sorted numeric vector
#'
#' @examples
#' selection_sort(c(5, 2, 8, 3, 1))
#'
#' @note
#' Time Complexity: O(n^2)
#' Space Complexity: O(1)
#' Best for small datasets
selection_sort <- function(arr) {
  n <- length(arr)
  
  if (n <= 1) {
    return(arr)
  }
  
  for (i in 1:(n - 1)) {
    min_idx <- find_min_index(arr, i, n)
    
    if (i != min_idx) {
      # Swap elements
      temp <- arr[i]
      arr[i] <- arr[min_idx]
      arr[min_idx] <- temp
    }
  }
  
  return(arr)
}

# ----------------------------
# ✅ Test Cases
# ----------------------------

# Edge Case: Empty vector
print(selection_sort(c()))  # Expected: c()

# Edge Case: Single element
print(selection_sort(c(42)))  # Expected: c(42)

# Normal Case
print(selection_sort(c(5, 2, 9, 1, 6)))  # Expected: c(1, 2, 5, 6, 9)

# Already sorted
print(selection_sort(c(1, 2, 3, 4)))  # Expected: c(1, 2, 3, 4)

# Reverse order
print(selection_sort(c(9, 7, 5, 3, 1)))  # Expected: c(1, 3, 5, 7, 9)

# With duplicates
print(selection_sort(c(4, 2, 4, 3, 1)))  # Expected: c(1, 2, 3, 4, 4)
